swarm/src/connection.rs

```diff
@@ -367,6 +430,15 @@ where
     }
 }
 
+fn gather_supported_protocols(handler: &impl ConnectionHandler) -> HashSet<String> {
+    handler
+        .listen_protocol()
+        .upgrade()
+        .protocol_info()
+        .filter_map(|i| String::from_utf8(i.protocol_name().to_vec()).ok())
```

Comment:
How about logging on `debug` in case this fails? Though I guess with https://github.com/libp2p/rust-libp2p/pull/3745/ this changes anyways.

Reply:
Yeah, let's see what happens once we have #3745 in.